### ISSUE TITLE
send -D -R (-I) regressions

### ISSUE THREAD

**Issue by kpande** [zfsonlinux/zfs#3066]
Sending from a recent Linux host (with kmem rework and latest master as of two days ago) with 'send -D -R -I' on a ZVOL with more than 2 snapshots will fail on FreeBSD and Linux recv "invalid backup stream"

https://gist.github.com/kpande/9ba7e73ea188113f473b

I can't ^C from the send process either. Server requires reboot or else ZFS commands fail.

**Comment by kpande** [comment]
probably related to #2210

**Comment by behlendorf** [comment]
Yes, this does look similar to #2210.  From your perspective is this a recent regression or just the first time you've tried this?

**Comment by kpande** [comment]
it is recent, though it seems to be an issue with the sending side. It occurs whether the recv is to FreeBSD or Linux.

I did have send -D working, in combination with -R -I, but sometime in the last month there was an update that broke it.

**Comment by kpande** [comment]
i did hit this bug on systems without the kmem rework.

**Comment by kpande** [comment]
@behlendorf this issue is still present in latest builds.

**Comment by mike503** [comment]
+1. locks it up too, cannot control-C out of it or kill -9 the process. have to reboot to clear it out.

dpkg -l | grep zfs:

ii  dkms                                2.2.0.3-1.1ubuntu5.14.10+zfs10+utopic
ii  libzfs2                             0.6.3-15~utopic
ii  mountall                            2.54ubuntu0.14.10.1-zfs1
ii  ubuntu-zfs                          9~utopic
ii  zfs-dkms                            0.6.3-15~utopic
ii  zfs-doc                             0.6.3-15~utopic
ii  zfsutils                            0.6.3-15~utopic

lsb_release -a:
No LSB modules are available.
Distributor ID: Ubuntu
Description:    Ubuntu 14.10
Release:        14.10
Codename:       utopic

uname -a:
Linux backup 3.16.0-31-generic #43-Ubuntu SMP Tue Mar 10 17:37:36 UTC 2015 x86_64 x86_64 x86_64 GNU/Linux

**Comment by baitisj** [comment]
I can confirm that this bug exists in FreeBSD 10.2.

**Comment by kpande** [comment]
@baitisj do you suppose then it came from an illumos commit?

**Comment by baitisj** [comment]
It's worth trying this on Illumos to see if it is affected, yes.

**Comment by behlendorf** [comment]
If this used to work and it's still broken in ZoL it would be useful to bisect which commit introduced this.

**Comment by ptx0** [comment]
I've been able to use send -D but haven't tested with a large dataset. It's dangerous to test so I haven't been able to get it safely.

**Comment by kpande** [comment]
Still hangs when doing ^C on `zfs send -D -R -v`

Stack traces: http://sprunge.us/FIQK

Process list: http://sprunge.us/CeTB

**Comment by gmelikov** [comment]
I think that #5689 may close this issue, reopen it again if it's not.

**Comment by kpande** [comment]
@gmelikov this is not resolved in 0.7, though the stack is different.

This hang is produced by `send -D`, on the receiving side.

```
INFO: task receive_writer:7442 blocked for more than 120 seconds.
      Tainted: P           O    4.4.63 #1
"echo 0 > /proc/sys/kernel/hung_task_timeout_secs" disables this message.
receive_writer  D ffff8823241d3d48     0  7442      2 0x00000000
 ffff8823241d3d48 ffff882359c18600 ffff8822f04d6900 ffff882300000000
 ffffffffa0681490 ffff8823241d3cb8 ffffffffa05a7fe7 ffff8822c32521e0
 ffff8822de760080 ffff8822de036050 0000000000000000 ffff8822de0360d0
Call Trace:
 [<ffffffffa05a7fe7>] ? dnode_rele_and_unlock+0x57/0xa0 [zfs]
 [<ffffffffa05a8063>] ? dnode_rele+0x33/0x40 [zfs]
 [<ffffffffa058b20e>] ? dbuf_rele_and_unlock+0x2de/0x390 [zfs]
 [<ffffffffa04c517b>] ? spl_kmem_free+0x2b/0x40 [spl]
 [<ffffffffa05a4dcc>] ? dmu_tx_destroy+0xfc/0x120 [zfs]
 [<ffffffff816c0b20>] schedule+0x40/0x90
 [<ffffffffa04ccaab>] cv_wait_common+0x11b/0x140 [spl]
 [<ffffffff810c39a0>] ? woken_wake_function+0x20/0x20
 [<ffffffffa058b97e>] ? dmu_buf_rele+0xe/0x10 [zfs]
 [<ffffffffa04ccb25>] __cv_wait+0x15/0x20 [spl]
 [<ffffffffa059011f>] bqueue_dequeue+0x3f/0xf0 [zfs]
 [<ffffffffa05a17e9>] receive_writer_thread+0x69/0x150 [zfs]
 [<ffffffffa05a1780>] ? dmu_recv_stream+0x7b0/0x7b0 [zfs]
 [<ffffffffa04c7b90>] ? __thread_create+0x150/0x150 [spl]
 [<ffffffffa04c7c08>] thread_generic_wrapper+0x78/0x90 [spl]
 [<ffffffffa04c7b90>] ? __thread_create+0x150/0x150 [spl]
 [<ffffffff810a2f84>] kthread+0xe4/0x100
 [<ffffffff810a2ea0>] ? kthread_freezable_should_stop+0x70/0x70
 [<ffffffff816c51cf>] ret_from_fork+0x3f/0x70
 [<ffffffff810a2ea0>] ? kthread_freezable_should_stop+0x70/0x70
```

**Comment by kpande** [comment]
memory is not a concern, but I have seen "unable to handle page" in kernel buffer
```
             total       used       free     shared    buffers     cached
Mem:        141035       4205     136830          0        126        426
-/+ buffers/cache:       3653     137382
Swap:         4095          0       4095
```
